File: src/Modules/Patient/Modules.Patient.Contracts/Authorization/PatientPermissions.cs
using FSH.Framework.Shared.Constants;

namespace FSH.Modules.Patient.Contracts.Authorization;

public static class PatientPermissions
{
    public static class Patients
    {
        public const string Resource = "Patient.Patients";
        public const string View    = $"Permissions.{Resource}.View";
        public const string Create  = $"Permissions.{Resource}.Create";
        public const string Update  = $"Permissions.{Resource}.Update";
        public const string Delete  = $"Permissions.{Resource}.Delete";
        public const string Restore = $"Permissions.{Resource}.Restore";
    }

    public static class Incidents
    {
        public const string Resource = "Patient.Incidents";
        public const string View   = $"Permissions.{Resource}.View";
        public const string Create = $"Permissions.{Resource}.Create";
        public const string Update = $"Permissions.{Resource}.Update";
        public const string Close  = $"Permissions.{Resource}.Close";
        public const string Delete = $"Permissions.{Resource}.Delete";
    }

    public static class Reports
    {
        public const string Resource = "Patient.Reports";
        public const string View   = $"Permissions.{Resource}.View";
        public const string Create = $"Permissions.{Resource}.Create";
        public const string Update = $"Permissions.{Resource}.Update";
        public const string Sign   = $"Permissions.{Resource}.Sign";
        public const string Review = $"Permissions.{Resource}.Review";
        public const string Delete = $"Permissions.{Resource}.Delete";
    }

    public static class Problems
    {
        public const string Resource = "Patient.Problems";
        public const string View   = $"Permissions.{Resource}.View";
        public const string Create = $"Permissions.{Resource}.Create";
        public const string Update = $"Permissions.{Resource}.Update";
        public const string Delete = $"Permissions.{Resource}.Delete";
    }

    public static class Allergies
    {
        public const string Resource = "Patient.Allergies";
        public const string View   = $"Permissions.{Resource}.View";
        public const string Create = $"Permissions.{Resource}.Create";
        public const string Update = $"Permissions.{Resource}.Update";
        public const string Delete = $"Permissions.{Resource}.Delete";
    }

    public static IReadOnlyList<FshPermission> All { get; } =
    [
        new("View Patients",    ActionConstants.View,   Patients.Resource, IsBasic: true),
        new("Create Patients",  ActionConstants.Create, Patients.Resource),
        new("Update Patients",  ActionConstants.Update, Patients.Resource),
        new("Delete Patients",  ActionConstants.Delete, Patients.Resource),
        new("Restore Patients", "Restore",              Patients.Resource),
        new("View Incidents",   ActionConstants.View,   Incidents.Resource, IsBasic: true),
        new("Create Incidents", ActionConstants.Create, Incidents.Resource),
        new("Update Incidents", ActionConstants.Update, Incidents.Resource),
        new("Close Incidents",  "Close",                Incidents.Resource),
        new("Delete Incidents", ActionConstants.Delete, Incidents.Resource),
        new("View Reports",   ActionConstants.View,   Reports.Resource, IsBasic: true),
        new("Create Reports", ActionConstants.Create, Reports.Resource),
        new("Update Reports", ActionConstants.Update, Reports.Resource),
        new("Sign Reports",   "Sign",                 Reports.Resource),
        new("Review Reports", "Review",               Reports.Resource),
        new("Delete Reports", ActionConstants.Delete, Reports.Resource),
        new("View Problems",   ActionConstants.View,   Problems.Resource, IsBasic: true),
        new("Create Problems", ActionConstants.Create, Problems.Resource),
        new("Update Problems", ActionConstants.Update, Problems.Resource),
        new("Delete Problems", ActionConstants.Delete, Problems.Resource),
        new("View Allergies",   ActionConstants.View,   Allergies.Resource, IsBasic: true),
        new("Create Allergies", ActionConstants.Create, Allergies.Resource),
        new("Update Allergies", ActionConstants.Update, Allergies.Resource),
        new("Delete Allergies", ActionConstants.Delete, Allergies.Resource),
    ];
}

File: clients/dashboard/src/lib/patient-permissions.ts
/**
 * Permissions for the Patient feature. Mirrors the server registry
 * (`PatientPermissions.Patients` in `Modules.Patient.Contracts`) — convention
 * follows `Permissions.{Resource}.{Action}`, same shape as
 * [[trash-permissions]]'s `TRASH_TAB_PERMISSIONS`. If the server permission
 * set changes, mirror it here.
 */
export const PATIENT_PERMISSIONS = {
  view: "Permissions.Patient.Patients.View",
  create: "Permissions.Patient.Patients.Create",
  update: "Permissions.Patient.Patients.Update",
  delete: "Permissions.Patient.Patients.Delete",
  restore: "Permissions.Patient.Patients.Restore",
} as const;

export type PatientPermissionKey = keyof typeof PATIENT_PERMISSIONS;

/** Flat list of every Patient permission. */
export const ALL_PATIENT_PERMISSIONS: readonly string[] = Object.values(PATIENT_PERMISSIONS);

export const INCIDENT_PERMISSIONS = {
  view:   "Permissions.Patient.Incidents.View",
  create: "Permissions.Patient.Incidents.Create",
  update: "Permissions.Patient.Incidents.Update",
  close:  "Permissions.Patient.Incidents.Close",
  delete: "Permissions.Patient.Incidents.Delete",
} as const;

export type IncidentPermissionKey = keyof typeof INCIDENT_PERMISSIONS;

export const REPORT_PERMISSIONS = {
  view:   "Permissions.Patient.Reports.View",
  create: "Permissions.Patient.Reports.Create",
  update: "Permissions.Patient.Reports.Update",
  sign:   "Permissions.Patient.Reports.Sign",
  review: "Permissions.Patient.Reports.Review",
  export: "Permissions.Patient.Reports.Export",
  delete: "Permissions.Patient.Reports.Delete",
} as const;

export type ReportPermissionKey = keyof typeof REPORT_PERMISSIONS;

export const PROBLEM_PERMISSIONS = {
  view:   "Permissions.Patient.Problems.View",
  create: "Permissions.Patient.Problems.Create",
  update: "Permissions.Patient.Problems.Update",
  delete: "Permissions.Patient.Problems.Delete",
} as const;

export type ProblemPermissionKey = keyof typeof PROBLEM_PERMISSIONS;

export const ALLERGY_PERMISSIONS = {
  view:   "Permissions.Patient.Allergies.View",
  create: "Permissions.Patient.Allergies.Create",
  update: "Permissions.Patient.Allergies.Update",
  delete: "Permissions.Patient.Allergies.Delete",
} as const;

export type AllergyPermissionKey = keyof typeof ALLERGY_PERMISSIONS;

export const MEDICATION_PERMISSIONS = {
  view:   "Permissions.Patient.Medications.View",
  create: "Permissions.Patient.Medications.Create",
  update: "Permissions.Patient.Medications.Update",
  delete: "Permissions.Patient.Medications.Delete",
} as const;

export type MedicationPermissionKey = keyof typeof MEDICATION_PERMISSIONS;

export const NOTE_PERMISSIONS = {
  view:   "Permissions.Patient.Notes.View",
  create: "Permissions.Patient.Notes.Create",
  update: "Permissions.Patient.Notes.Update",
  delete: "Permissions.Patient.Notes.Delete",
} as const;

export type NotePermissionKey = keyof typeof NOTE_PERMISSIONS;

export const DOCUMENT_PERMISSIONS = {
  view:     "Permissions.Patient.Documents.View",
  create:   "Permissions.Patient.Documents.Create",
  update:   "Permissions.Patient.Documents.Update",
  download: "Permissions.Patient.Documents.Download",
  delete:   "Permissions.Patient.Documents.Delete",
} as const;

export type DocumentPermissionKey = keyof typeof DOCUMENT_PERMISSIONS;

export const SUPERBILL_PERMISSIONS = {
  view:   "Permissions.Patient.SuperBills.View",
  manage: "Permissions.Patient.SuperBills.Manage",
} as const;

export type SuperBillPermissionKey = keyof typeof SUPERBILL_PERMISSIONS;
